lightning/src/offers/invoice.rs

```diff
@@ -416,32 +525,38 @@ impl UnsignedBolt12Invoice {
 	pub fn tagged_hash(&self) -> &TaggedHash {
 		&self.tagged_hash
 	}
+}
 
+macro_rules! unsigned_invoice_sign_method { ($self: ident, $self_type: ty) => {
 	/// Signs the [`TaggedHash`] of the invoice using the given function.
 	///
 	/// Note: The hash computation may have included unknown, odd TLV records.
 	///
 	/// This is not exported to bindings users as functions aren't currently mapped.
-	pub fn sign<F, E>(mut self, sign: F) -> Result<Bolt12Invoice, SignError<E>>
+	pub fn sign<F, E>(mut $self: $self_type, sign: F) -> Result<Bolt12Invoice, SignError<E>>
```

Comment:
Oh, yeah... I may have either been overzealous or meant to complete this lol. I guess we can make the `Fn` bound use a trait instead and implement it for all functions? I seem to recall awhile ago this was requiring me to specify the type when calling this method in tests. Or maybe that was with a type alias? Anyhow, I can try changing it to see what happens.

Reply:
Latest push uses a `SignFunction` trait instead of `Fn`.

Reply:
For the method, the compiler doesn't like defaults:

```
error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions
   --> lightning/src/offers/invoice_request.rs:517:17
    |
517 |     pub fn sign<F, E = ()>($($self_mut)* $self: $self_type, sign: F) -> Result<InvoiceRequest, SignError<E>>
    |                    ^^^^^^
...
546 |     unsigned_invoice_request_sign_method!(self, Self, mut);
    |     ------------------------------------------------------ in this macro invocation
    |
    = note: `#[deny(invalid_type_param_default)]` on by default
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
    = note: this error originates in the macro `unsigned_invoice_request_sign_method` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `lightning` due to previous error
```

What if I made `E` an associated type on `SignFunction`? Compiler seems happy with this.

Or bound `E` by `core::error::Error`? Though this doesn't seem to be available in `core` using 1.63.0.

Reply:
Latest push uses a default parameter type in the enum and an associated type in the trait if you want to try it.